presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java

```diff
@@ -145,6 +226,21 @@ private synchronized PrestoNode refreshNodesInternal()
         return currentNode;
     }
 
+    private boolean isNodeShuttingDown(String nodeId)
+    {
+        // check shutdownFlag for the current node's state
+        // for other nodes check nodeStates map (which is non-empty only on coordinator)
+        if (nodeId.equals(nodeInfo.getNodeId())) {
+            return shutdownFlag.get();
```

Comment:
I don't think you need this `shutdownFlag`. The coordinator should be able to talk to itself to retrieve it's own state, and then it looks like you can remove the whole `ShutdownFlag` class